### PR TITLE
refactor(tree): Misc. column/row insert/remove improvements

### PR DESCRIPTION
**Pull request by Josmithr**
There is a significant perf difference between inserting single columns/rows and inserting lists of them. To help encourage group inserts, `insertColumn` and `insertRow` have been removed. Users can leverage the group alternatives instead (`insertColumns` and `insertRows`).

Additionally, there is a significant performance difference between _removing_ single / disjoint columns/rows and removing contiguous ranges of them. To address this, overloads have been added to `removeColumns` and `removeRows` to allow removing by index range. When possible, this is the most performant option. `removeColumns` and `removeRows` have also been updated to calculate the set of contiguous ranges represented lists of nodes or IDs to minimize the number of removal ops made in each transaction.

### Benchmark test updates

Updates both SharedTree table and SharedMatrix perf tests to include batch column/row insertion tests.

Batch removal tests will come in a future PR, due to some of the current limitations around column + cell removal.

### CODE REVIEW

File: packages/dds/tree/api-report/tree.alpha.api.md [hunk 1]
```diff
@@ -1150,18 +1150,10 @@ export namespace TableSchema {
     export function column<const TScope extends string | undefined, const TCell extends ImplicitAllowedTypes, const TProps extends ImplicitAnnotatedFieldSchema>(params: System_TableSchema.CreateColumnOptionsBase<SchemaFactoryAlpha<TScope>, TCell> & {
         readonly props: TProps;
     }): System_TableSchema.ColumnSchemaBase<TScope, TCell, TProps>;
```
Comment (Josmithr) [inline packages/dds/tree/api-report/tree.alpha.api.md]: TODO: add changeset

**Review by WayneFerrao — COMMENTED** [review]
Looks good at first glance. Makes sense that bulk updates are performant but It'd be cool to see stats/ run times which show the difference before and after these changes.

File: packages/dds/tree/src/tableSchema.ts [hunk 2]
```diff
@@ -745,96 +723,178 @@ export namespace System_TableSchema {
...
+
 				Tree.runTransaction(this, () => {
 					// Note, throwing an error within a transaction will abort the entire transaction.
-					// So if we throw an error here for any row, no columns will be removed.
-					for (const columnToRemove of columns) {
-						const removedRow = this.removeColumn(columnToRemove);
-						removedColumns.push(removedRow);
+					// So if we throw an error here for any column, no columns will be removed.
+					for (const range of rangesToRemove) {
+						this.removeRangeOfColumns(range.start, range.end);
```
Comment (Josmithr) [inline packages/dds/tree/src/tableSchema.ts]: TODO: this is wrong (for both columns and rows). Each iteration we need to adjust the remaining ranges to account for the previously removed items

File: packages/dds/matrix/src/test/memory/sharedMatrix.spec.ts [hunk 3]
```diff
@@ -284,7 +284,7 @@ describe("SharedMatrix memory usage", () => {
 			// Insert-related tests that are not limited by matrixSize
 			for (const count of operationCounts) {
 				describe(`Column Insertion`, () => {
-					// Test the memory usage of the SharedMatrix for inserting a column in the middle for a given number of times.
+					// Test the memory usage of inserting a single column in the middle of the table N times.
```
Comment (jzaffiro) [inline packages/dds/matrix/src/test/memory/sharedMatrix.spec.ts]:
```suggestion
					// Test the memory usage of inserting a single empty column in the middle of the table N times.
```
Should the comments above the column benchmarks be updated to match those of the row benchmarks? (feel free to ignore if the column is not in fact empty)

File: packages/dds/matrix/src/test/time/sharedMatrix.spec.ts [hunk 4]
```diff
@@ -581,7 +683,7 @@ describe("SharedMatrix execution time", () => {
 						maxBenchmarkDurationSeconds,
 					});
 
-					// Test the execute time of the SharedMatrix for inserting a row and a column and removing them right away for a given number of times.
+					// Test the execution time of inserting a single empty row and a column and removing them right away N times.
```
Comment (jzaffiro) [inline packages/dds/matrix/src/test/time/sharedMatrix.spec.ts]: Overall, the use of "empty" seems inconsistent across the various tests, and I think it would be good to clear up what each test is inserting/undoing/redoing.

File: packages/dds/matrix/src/test/memory/sharedMatrix.spec.ts (hunk 3 above)
Comment (Josmithr) [reply packages/dds/matrix/src/test/memory/sharedMatrix.spec.ts]: There actually is a distinction here, which is the reason the docs are a bit different. In the SharedTree table model, the `row` objects hold the list of cells. The `column` objects are really just glorified IDs (with some optional props). So an "empty" column doesn't really have much meaning, but an "empty" row does. In the future, I'm planning to add variants of the row insertion tests that insert "non-empty" rows so we can compare the 2 scenarios.

File: packages/dds/matrix/src/test/time/sharedMatrix.spec.ts (hunk 4 above)
Comment (Josmithr) [reply packages/dds/matrix/src/test/time/sharedMatrix.spec.ts]: I'll go make sure all of the row insertion cases call out that the rows being inserted are empty.

**Comment by Josmithr** [comment]
Waiting on #25213

File: packages/dds/matrix/src/test/matrixBenchmarks.ts [hunk 5]
```diff
@@ -269,6 +269,8 @@ function runBenchmark(options: BenchmarkOptions): Test {
 	}
 }
 
+// TODO: add range-based insertion and removal tests
+
```
Comment (Josmithr) [inline packages/dds/matrix/src/test/matrixBenchmarks.ts:273]: Already done

Comment (Josmithr) [inline packages/dds/matrix/src/test/matrixBenchmarks.ts:273]:
```suggestion
```

File: packages/dds/tree/src/test/memory/tableTree.spec.ts [hunk 6]
```diff
@@ -248,6 +248,59 @@ describe("SharedTree table APIs memory usage", () => {
...
+
+						undoRedoManager.undo();
+						assert(!undoRedoManager.canUndo);
+						assert(undoRedoManager.canRedo);
+					},
+					operation: (table, undoRedoManager) => {
+						undoRedoManager.redo();
+						assert(!undoRedoManager.canRedo);
+					},
+				});
```
Comment (Josmithr) [inline packages/dds/tree/src/test/memory/tableTree.spec.ts:303]:
```suggestion
```

**Comment by Josmithr** [comment]
All of the test changes here have been merged in separately, as well as some of the structural changes in the table schema code. The final perf improvement is now being done in a separate PR here: https://github.com/microsoft/FluidFramework/pull/25329

Closing this PR.
